File: Objective-C/Parrot/Parrot.m
#import "Parrot.h"

@implementation Parrot {
    ParrotTypeEnum _type;
    NSInteger _numberOfCoconuts;
    double _voltage;
    BOOL _nailed;
}

- (instancetype)initWithType:(ParrotTypeEnum)type numberOfCoconuts:(NSInteger)numberOfCoconuts voltage:(double)voltage nailed:(BOOL)nailed {
    if ((self = [super init])) {
        _type = type;
        _numberOfCoconuts = numberOfCoconuts;
        _voltage = voltage;
        _nailed = nailed;
    }
    return self;
}

- (double)speed {
    switch(_type) {
        case ParrotTypeEnumEuropean:
            return self.baseSpeed;
            
        case ParrotTypeEnumAfrican:
            return MAX(0.0, self.baseSpeed - self.loadFactor * _numberOfCoconuts);
            
        case ParrotTypeEnumNorwegianBlue:
            return (_nailed) ? 0 : [self baseSpeed:_voltage];
    }
    @throw [NSException exceptionWithName:@"Should be unreachable" reason:nil userInfo:nil];
}

- (double)baseSpeed:(double)voltage {
    return MIN(24.0, voltage*self.baseSpeed);
}

- (double)loadFactor {
    return 9.0;
}

- (double)baseSpeed {
    return 12.0;
}

@end

File: Go/parrot.go
package parrot

import (
	"errors"
	"math"
)

type parrotType int

const (
	TypeEuropean      parrotType = 1
	TypeAfrican       parrotType = 2
	TypeNorwegianBlue parrotType = 3
)

// Parrot has a Speed and Cry.
type Parrot interface {
	Speed() (float64, error)
	Cry() (string, error)
}

type mixedParrot struct {
	_type            parrotType
	numberOfCoconuts int
	voltage          float64
	nailed           bool
}

func CreateParrot(t parrotType, numberOfCoconuts int, voltage float64, nailed bool) Parrot {
	return mixedParrot{t, numberOfCoconuts, voltage, nailed}
}

func (parrot mixedParrot) Speed() (float64, error) {
	switch parrot._type {
	case TypeEuropean:
		return parrot.baseSpeed(), nil

	case TypeAfrican:
		return math.Max(0, parrot.baseSpeed()-parrot.loadFactor()*float64(parrot.numberOfCoconuts)), nil

	case TypeNorwegianBlue:
		if parrot.nailed {
			return 0, nil
		}
		return parrot.computeBaseSpeedForVoltage(parrot.voltage), nil

	default:
		return 0, errors.New("should be unreachable")
	}
}

func (parrot mixedParrot) computeBaseSpeedForVoltage(voltage float64) float64 {
	return math.Min(24.0, voltage*parrot.baseSpeed())
}

func (parrot mixedParrot) loadFactor() float64 {
	return 9.0
}

func (parrot mixedParrot) baseSpeed() float64 {
	return 12.0
}

func (parrot mixedParrot) Cry() (string, error) {
	switch parrot._type {
	case TypeEuropean:
		return "Sqoork!", nil

	case TypeAfrican:
		return "Sqaark!", nil

	case TypeNorwegianBlue:
		if parrot.voltage > 0 {
			return "Bzzzzzz", nil
		}
		return "...", nil

	default:
		return "", errors.New("should be unreachable")
	}
}
